File: 3BIT/IIS/IIS-Backend/Postman/InitDB.js
console.log("Initialization Start");

const auditoriumsData = [
    { name: "D105", capacity: 300 },
    { name: "D206", capacity: 40 },
    { name: "D207", capacity: 100 },
    { name: "B101", capacity: 50 },
    { name: "A304", capacity: 150 },
];

const usersData = [
    { name: "Alice Johnson", email: "alice.johnson@example.com", role: "admin", picture: "https://example.com/pictures/alice.jpg" },
    { name: "Bob Smith", email: "bob.smith@example.com", role: "user", picture: "https://example.com/pictures/bob.jpg" },
    { name: "Charlie Brown", email: "charlie.brown@example.com", role: "user", picture: "https://example.com/pictures/charlie.jpg" },
    { name: "Dana White", email: "dana.white@example.com", role: "admin", picture: "https://example.com/pictures/dana.jpg" },
    { name: "Evan Green", email: "evan.green@example.com", role: "user", picture: "https://example.com/pictures/evan.jpg" },
];

const conferencesData = [
    { 
        orginizer_id: 1, 
        name: "Tech Innovations 2024", 
        description: "Annual conference on the latest in technology and innovation.", 
        capacity: 500, 
        cost: 150, 
        max_reservations_per_user: 3, 
        current_occupancy: 350, 
        picture: "https://example.com/pictures/tech_innovations.jpg", 
        tags: ["technology", "innovation", "2024"], 
        start_time: "2024-05-15T09:00:00Z", 
        end_time: "2024-05-15T17:00:00Z" 
    },
    // { 
    //     orginizer_id: 2, 
    //     name: "Health & Wellness Expo", 
    //     description: "Explore new trends in health and wellness.", 
    //     capacity: 300, 
    //     cost: 50, 
    //     max_reservations_per_user: 2, 
    //     current_occupancy: 200, 
    //     picture: "https://example.com/pictures/health_expo.jpg", 
    //     tags: ["health", "wellness", "expo"], 
    //     start_time: "2024-06-20T10:00:00Z", 
    //     end_time: "2024-06-20T18:00:00Z" 
    // },
    // { 
    //     orginizer_id: 3, 
    //     name: "Green Energy Summit", 
    //     description: "A gathering of industry leaders to discuss renewable energy.", 
    //     capacity: 400, 
    //     cost: 100, 
    //     max_reservations_per_user: 4, 
    //     current_occupancy: 150, 
    //     picture: "https://example.com/pictures/green_energy.jpg", 
    //     tags: ["energy", "green", "summit"], 
    //     start_time: "2024-07-10T08:00:00Z", 
    //     end_time: "2024-07-10T16:00:00Z" 
    // },
    // { 
    //     orginizer_id: 4, 
    //     name: "Digital Marketing World", 
    //     description: "Insights into the world of digital marketing.", 
    //     capacity: 250, 
    //     cost: 75, 
    //     max_reservations_per_user: 2, 
    //     current_occupancy: 125, 
    //     picture: "https://example.com/pictures/marketing_world.jpg", 
    //     tags: ["digital marketing", "marketing", "2024"], 
    //     start_time: "2024-08-05T09:30:00Z", 
    //     end_time: "2024-08-05T17:00:00Z" 
    // },
    // { 
    //     orginizer_id: 5, 
    //     name: "Artificial Intelligence Symposium", 
    //     description: "Deep dive into AI and machine learning trends.", 
    //     capacity: 600, 
    //     cost: 200, 
    //     max_reservations_per_user: 5, 
    //     current_occupancy: 450, 
    //     picture: "https://example.com/pictures/ai_symposium.jpg", 
    //     tags: ["AI", "machine learning", "symposium"], 
    //     start_time: "2024-09-12T09:00:00Z", 
    //     end_time: "2024-09-12T18:00:00Z" 
    // }
];

const presentationsData = [
    { 
        speaker_id: 2, 
        conference_id: 1, 
        name: "The Future of AI", 
        description: "Exploring advancements and ethical considerations in artificial intelligence.", 
        picture: "https://example.com/pictures/ai_future.jpg", 
        presentation_text: "In this talk, we will dive into the latest breakthroughs in AI...", 
        auditorium_id: 1, 
        status: "scheduled", 
        start_time: "2024-05-15T10:00:00Z", 
        end_time: "2024-05-15T11:30:00Z" 
    },
    // { 
    //     speaker_id: 2, 
    //     conference_id: 2, 
    //     name: "Wellness in the Workplace", 
    //     description: "Strategies to promote health and wellness among employees.", 
    //     picture: "https://example.com/pictures/wellness_workplace.jpg", 
    //     presentation_text: "This session covers effective wellness programs and initiatives...", 
    //     auditorium_id: 102, 
    //     status: "scheduled", 
    //     start_time: "2024-06-20T11:00:00Z", 
    //     end_time: "2024-06-20T12:30:00Z" 
    // },
    // { 
    //     speaker_id: 3, 
    //     conference_id: 3, 
    //     name: "Renewable Energy Challenges", 
    //     description: "Discussing barriers and solutions in renewable energy adoption.", 
    //     picture: "https://example.com/pictures/energy_challenges.jpg", 
    //     presentation_text: "This talk will examine the key obstacles in renewable energy...", 
    //     auditorium_id: 103, 
    //     status: "scheduled", 
    //     start_time: "2024-07-10T14:00:00Z", 
    //     end_time: "2024-07-10T15:30:00Z" 
    // },
    // { 
    //     speaker_id: 4, 
    //     conference_id: 4, 
    //     name: "Social Media Strategies", 
    //     description: "Effective techniques for digital marketing through social media.", 
    //     picture: "https://example.com/pictures/social_media.jpg", 
    //     presentation_text: "In today's session, we'll explore ways to leverage social media...", 
    //     auditorium_id: 104, 
    //     status: "scheduled", 
    //     start_time: "2024-08-05T13:00:00Z", 
    //     end_time: "2024-08-05T14:30:00Z" 
    // },
    // { 
    //     speaker_id: 5, 
    //     conference_id: 5, 
    //     name: "Machine Learning in Healthcare", 
    //     description: "How machine learning is transforming the healthcare industry.", 
    //     picture: "https://example.com/pictures/ml_healthcare.jpg", 
    //     presentation_text: "This talk focuses on machine learning applications in healthcare...", 
    //     auditorium_id: 105, 
    //     status: "scheduled", 
    //     start_time: "2024-09-12T09:30:00Z", 
    //     end_time: "2024-09-12T11:00:00Z" 
    // }
];

const reservationsData = [
    { 
        conference_id: 1, 
        user_id: 3, 
        visitors_count: 2, 
        status: "confirmed" 
    },
    // { 
    //     conference_id: 2, 
    //     user_id: 2, 
    //     visitors_count: 1, 
    //     status: "pending" 
    // },
    // { 
    //     conference_id: 3, 
    //     user_id: 3, 
    //     visitors_count: 4, 
    //     status: "cancelled" 
    // },
    // { 
    //     conference_id: 4, 
    //     user_id: 4, 
    //     visitors_count: 3, 
    //     status: "confirmed" 
    // },
    // { 
    //     conference_id: 5, 
    //     user_id: 5, 
    //     visitors_count: 1, 
    //     status: "pending" 
    // }
];


function createData(data, api_type) {
    pm.sendRequest({
        url: pm.collectionVariables.get("base_url") + "/api/" + api_type,
        method: "POST",
        header: { "Content-Type": "application/json" },
        body: {
            mode: "raw",
            raw: JSON.stringify(data)
        }
    }, (err, res) => {
        if (err) {
            console.log("Error while creating: ", err);
        } else {
            console.log("Created: for "+api_type, res.json());
        }
    });
}

function checkData(data, api_type) {

    pm.sendRequest(pm.collectionVariables.get("base_url") + "/api/" + api_type, function (err, res) {
        if (err) {
            console.log("Error while checking: " + api_type, err);
            return;
        }
        const countAfterDelete = res.json().length;
        pm.test("All " + api_type + " was created successfully", function () {
            pm.expect(countAfterDelete).to.equal(data.length);
        });
    });
}

auditoriumsData.forEach(data => createData(data, "auditoriums"));
checkData(auditoriumsData, "auditoriums");

usersData.forEach(data => createData(data, "users"));
checkData(usersData, "users");

conferencesData.forEach(data => createData(data, "conferences"));
checkData(conferencesData, "conferences");

presentationsData.forEach(data => createData(data, "presentations"));
checkData(presentationsData, "presentations");

reservationsData.forEach(data => createData(data, "reservations"));
checkData(reservationsData, "reservations");

console.log("Initialization End");

File: 3BIT/IIS/IIS-Backend/init_db.py
# Import necessary models and datetime
from db import db
from datetime import datetime
from db.auditorium_model import Auditorium
from db.ConferenceAuditorium import ConferenceAuditorium
from db.user_model import User
from db.conference_model import Conference
from db.presentation_model import Presentation
from db.reservation_model import Reservation
from werkzeug.security import generate_password_hash


def populate_db():
    """Populate the database with initial data."""

    print("Deleting existing data...")

    # Delete all records from dependent tables first to avoid foreign key constraint issues
    ConferenceAuditorium.query.delete()
    Reservation.query.delete()
    Presentation.query.delete()
    Conference.query.delete()
    Auditorium.query.delete()
    User.query.delete()
    db.session.commit()

    print("Populating new data...")

    # Populate Auditoriums
    auditoriums_data = [
        {"name": "D105", "capacity": 300},
        {"name": "D206", "capacity": 40},
        {"name": "D207", "capacity": 100},
        {"name": "B101", "capacity": 50},
        {"name": "A304", "capacity": 150},
    ]
    auditoriums = [Auditorium(**data) for data in auditoriums_data]
    db.session.add_all(auditoriums)
    db.session.commit()

    # Populate Users with password hashing
    users_data = [
        {"name": "Alice Johnson", "email": "alice.johnson@example.com", "role": "user", "picture": "https://static.vecteezy.com/system/resources/thumbnails/038/974/578/small_2x/ai-generated-professional-portrait-of-a-competent-woman-free-photo.jpg", "password": "alice123"},
        {"name": "Bob Smith", "email": "bob.smith@example.com", "role": "user", "picture": "https://htmlstream.com/preview/unify-v2.6.2/assets/img-temp/400x450/img5.jpg", "password": "bob123"},
        {"name": "Charlie Brown", "email": "charlie.brown@example.com", "role": "user", "picture": "https://preview.keenthemes.com/metronic-v4/theme/assets/pages/media/profile/people19.png", "password": "charlie123"},
        {"name": "Dana White", "email": "dana.white@example.com", "role": "admin", "picture": "https://variety.com/wp-content/uploads/2018/10/dana_white.png", "password": "dana123"},
        {"name": "Evan Green", "email": "evan.green@example.com", "role": "user", "picture": "https://t3.ftcdn.net/jpg/04/60/91/88/360_F_460918802_XVCymFr7MoziFpnInbTDvrlblYhvAOi2.jpg", "password": "evan123"},
    ]

    for i in range(6, 21):
        users_data.append({
            "name": f"User{i}",
            "email": f"user{i}@example.com",
            "role": "user",
            "picture": f"https://example.com/pictures/user{i}.jpg",
            "password": f"user{i}123"
        })

    users = []
    for data in users_data:
        password_hash = generate_password_hash(data["password"])
        user = User(
            name=data["name"],
            email=data["email"],
            role=data["role"],
            picture=data["picture"],
            password_hash=password_hash
        )
        users.append(user)

    db.session.add_all(users)
    db.session.commit()

    # Now we can safely refer to users when adding conferences
    alice = User.query.filter_by(email="alice.johnson@example.com").first()
    bob = User.query.filter_by(email="bob.smith@example.com").first()
    charlie = User.query.filter_by(email="charlie.brown@example.com").first()

    # Populate Conferences
    conferences_data = [
        {
            "organizer_id": alice.id,
            "name": "Tech Innovations 2024",
            "description": "Annual conference on the latest in technology and innovation.",
            "capacity": 10,
            "cost": 150,
            "max_reservations_per_user": 3,
            "current_occupancy": 0,
            "picture": "https://weareprimegroup.com/_nuxt/image/f69e30.jpeg",
            "tags": ["technology", "innovation", "2024"],
            "start_time": datetime(2025, 1, 10, 9, 0),
            "end_time": datetime(2025, 1, 10, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "Old Conference",
            "description": "Some old conference. example",
            "capacity": 10,
            "cost": 150,
            "max_reservations_per_user": 3,
            "current_occupancy": 0,
            "picture": "https://images.theconversation.com/files/393210/original/file-20210401-13-z6rl6z.jpg?ixlib=rb-4.1.0&rect=0%2C292%2C3008%2C1504&q=45&auto=format&w=1356&h=668&fit=crop",
            "tags": ['innovation'],
            "start_time": datetime(2024, 1, 10, 9, 0),
            "end_time": datetime(2024, 1, 10, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "Healthcare Advances 2024",
            "description": "Exploring the latest trends in healthcare technology.",
            "capacity": 10,
            "cost": 100,
            "max_reservations_per_user": 2,
            "current_occupancy": 0,
            "picture": "https://assets.contenthub.wolterskluwer.com/api/public/content/2336569-female-doctor-using-a-digital-tablet-in-a-modern-hospital-16a7c0e8ea?v=dc468bea",
            "tags": ["healthcare", "innovation"],
            "start_time": datetime(2025, 1, 15, 9, 0),
            "end_time": datetime(2025, 1, 15, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "Green Energy Summit",
            "description": "Conference on renewable energy sources and sustainability.",
            "capacity": 30,
            "cost": 120,
            "max_reservations_per_user": 2,
            "current_occupancy": 0,
            "picture": "https://www.twi-global.com/CachedImage.axd?ImageName=Wind-Turbines.jpg&ImageWidth=784&ImageHeight=379&ImageVersionID=98786&ImageModified=20200624095115",
            "tags": ["energy", "sustainability"],
            "start_time": datetime(2025, 2, 5, 9, 0),
            "end_time": datetime(2025, 2, 5, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "AI and Ethics",
            "description": "Discussion on the ethical implications of artificial intelligence.",
            "capacity": 20,
            "cost": 80,
            "max_reservations_per_user": 1,
            "current_occupancy": 0,
            "picture": "https://eu-images.contentstack.com/v3/assets/blt6b0f74e5591baa03/blt7ff31ef3cc9e7065/65ae97402bb7f9040a716275/News_Image_-_2024-01-22T102627.569.png?disable=upscale&width=1200&height=630&fit=crop",
            "tags": ["ai", "ethics"],
            "start_time": datetime(2025, 3, 10, 9, 0),
            "end_time": datetime(2025, 3, 10, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "Blockchain for Business",
            "description": "How blockchain technology can be applied to business scenarios.",
            "capacity": 25,
            "cost": 200,
            "max_reservations_per_user": 2,
            "current_occupancy": 0,
            "picture": "https://cdn.media.amplience.net/i/epammarketplace/blockchain-uses-cryptocurrency?maxW=1200&qlt=80&fmt=jpg&bg=rgb(255,255,255)&v=126375874593",
            "tags": ["blockchain", "business"],
            "start_time": datetime(2025, 4, 20, 9, 0),
            "end_time": datetime(2025, 4, 20, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "Virtual Reality in Education",
            "description": "The impact of VR technologies in the education sector.",
            "capacity": 15,
            "cost": 50,
            "max_reservations_per_user": 1,
            "current_occupancy": 0,
            "picture": "https://www.computeam.co.uk/img/dynamic/1000/w/90/what-is-the-future-of-virtual-reality-in-education-34363342.jpg",
            "tags": ["vr", "education"],
            "start_time": datetime(2025, 5, 25, 9, 0),
            "end_time": datetime(2025, 5, 25, 17, 0),
        },
        {
            "organizer_id": alice.id,
            "name": "Space Exploration Forum",
            "description": "A deep dive into recent advancements in space exploration.",
            "capacity": 40,
            "cost": 180,
            "max_reservations_per_user": 2,
            "current_occupancy": 0,
            "picture": "https://res.cloudinary.com/momentum-media-group-pty-ltd/image/upload/v1686795211/Space%20Connect/space-exploration-sc_fm1ysf.jpg",
            "tags": ["space", "exploration"],
            "start_time": datetime(2025, 6, 5, 9, 0),
            "end_time": datetime(2025, 6, 5, 17, 0),
        },
    ]
    conferences = [Conference(**data) for data in conferences_data]
    db.session.add_all(conferences)
    db.session.commit()

    # Populate Reservations
    users = User.query.all()
    tech_conference = Conference.query.filter_by(name="Tech Innovations 2024").first()
    healthcare_conference = Conference.query.filter_by(name="Healthcare Advances 2024").first()

    # Create reservations for "Tech Innovations 2024" to make it sold out
    reservations_data = []
    for i in range(10):
        reservations_data.append({
            "conference_id": healthcare_conference.id,
            "user_id": users[i].id,
            "visitors_count": 1,
            "status": "approved",
        })

    reservations_data.append({
        "conference_id": tech_conference.id,
        "user_id": charlie.id,
        "visitors_count": 2,
        "status": "approved",
    })

    reservations = [Reservation(**data) for data in reservations_data]
    db.session.add_all(reservations)
    db.session.commit()

    # Update conference occupancy based on actual reservations
    tech_conference.current_occupancy = len([r for r in reservations if r.conference_id == tech_conference.id])
    healthcare_conference.current_occupancy = len([r for r in reservations if r.conference_id == healthcare_conference.id])
    db.session.commit()

    # Populate Presentations
    presentations_data = [
        {
            "speaker_id": bob.id,
            "conference_id": tech_conference.id,
            "name": "The Future of AI",
            "description": "Exploring advancements and ethical considerations in artificial intelligence.",
            "picture": "https://imageio.forbes.com/specials-images/imageserve/66bee357cf48b97789cbc270/0x0.jpg?format=jpg&height=900&width=1600&fit=bounds",
            "presentation_text": "In this talk, we will dive into the latest breakthroughs in AI...",
            "auditorium_id": auditoriums[0].id,
            "status": "confirmed",
            "start_time": datetime(2025, 1, 10, 10, 0),
            "end_time": datetime(2025, 1, 10, 11, 0),
        }
    ]
    presentations = [Presentation(**data) for data in presentations_data]
    db.session.add_all(presentations)
    db.session.commit()

    # Link Conferences to Auditoriums
    conference_auditorium_data = [
        {
            "conference_id": tech_conference.id,
            "auditorium_id": auditoriums[0].id
        },
        {
            "conference_id": healthcare_conference.id,
            "auditorium_id": auditoriums[1].id
        }
    ]
    conference_auditoriums = [ConferenceAuditorium(**data) for data in conference_auditorium_data]
    db.session.add_all(conference_auditoriums)
    db.session.commit()

    print("Database initialized successfully.")
